spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/BaseCatalog.java

```diff
@@ -46,4 +54,56 @@ public Procedure loadProcedure(Identifier ident) throws NoSuchProcedureException
 
     throw new NoSuchProcedureException(ident);
   }
+
+  @Override
+  public Identifier[] listFunctions(String[] namespace) throws NoSuchNamespaceException {
+    if (isValidNamespaceForFunction(namespace)) {
+      return Arrays.stream(SparkFunctions.list())
+          .map(name -> Identifier.of(namespace, name))
+          .toArray(Identifier[]::new);
+    } else {
+      throw new NoSuchNamespaceException(namespace);
+    }
+  }
+
+  @Override
+  public UnboundFunction loadFunction(Identifier ident) throws NoSuchFunctionException {
+    String[] namespace = ident.namespace();
+    String name = ident.name();
+
+    if (isValidNamespaceForFunction(namespace)) {
+      UnboundFunction func = SparkFunctions.load(name);
+      if (func != null) {
+        return func;
+      }
+    }
+
+    throw new NoSuchFunctionException(ident);
+  }
+
+  /**
+   * When in an Iceberg catalog, allow for using the built-in Iceberg functions provided that:
+   * <ol>
+   *   <li>
+   *     The namespace is not specified,
+   *     i.e. {@code SELECT truncate(1, 4)} or {@code SELECT my_catalog.truncate(1, 4)}
+   *   </li>
+   *   <li>
+   *     The implicit <b>system</b> namespace is used, to mirror call procedure syntax,
+   *     i.e. {@code SELECT system.truncate(1, 4)} or {@code SELECT my_catalog.system.truncate(1, 4)}
+   *   </li>
+   *   <li>
+   *     A namespace that exists within the catalog is referenced,
+   *     i.e. {@code SELECT ns.truncate(1, 4)} or {@code SELECT my_catalog.ns.truncate(1, 4)}
```

Comment:
Similar to my feedback on the Flink PR, this should not expose functions in every database. This should choose a database where Iceberg functions will be exposed. I recommend `system` since that's what we use for stored procedures.

Reply:
So the only problem with this is that within SQL (or even the dataframe API), the `SparkSessionCatalog` checks that the namespace exists.

So calling `system.truncate` results in:

Reply:
So the only problem with this is that within SQL (or even the dataframe API), the `SparkSessionCatalog` checks that the namespace exists.

So calling `system.truncate` results in
```
org.apache.spark.sql.AnalysisException: Undefined function: 'truncate'. This function is neither a registered temporary function nor a permanent function registered in the database 'system'.
```

I'm seeing if I can find a way around this, but it is part of the SessionCatalog's code.

Reply:
Full stack trace:

```
Undefined function: 'truncate'. This function is neither a registered temporary function nor a permanent function registered in the database 'system'.; line 1 pos 7
org.apache.spark.sql.AnalysisException: Undefined function: 'truncate'. This function is neither a registered temporary function nor a permanent function registered in the database 'system'.; line 1 pos 7
	at app//org.apache.spark.sql.catalyst.catalog.SessionCatalog.failFunctionLookup(SessionCatalog.scala:1561)
	at app//org.apache.spark.sql.catalyst.catalog.SessionCatalog.resolvePersistentFunctionInternal(SessionCatalog.scala:1704)
	at app//org.apache.spark.sql.catalyst.catalog.SessionCatalog.resolvePersistentFunction(SessionCatalog.scala:1673)
	at app//org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveFunctions$.org$apache$spark$sql$catalyst$analysis$Analyzer$ResolveFunctions$$resolveV1Function(Analyzer.scala:2172)
	at app//org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveFunctions$$anonfun$apply$25$$anonfun$applyOrElse$103.$anonfun$applyOrElse$109(Analyzer.scala:2122)
	at app//scala.Option.getOrElse(Option.scala:189)
	at app//org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveFunctions$$anonfun$apply$25$$anonfun$applyOrElse$103.$anonfun$applyOrElse$108(Analyzer.scala:2119)
```

Link to the code in the analyzer that's being hit: https://github.com/apache/spark/blob/f74867bddfbcdd4d08076db36851e88b15e66556/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L2118-L2126

Reply:
We spoke offline and we're just giong to hold off on the spark session catalog for now.

Most likely in a follow up we will add a namespace `system` that will then be resolvable (as these issues go away when the namespace is resolvable). That matches other databases, as well as the experience of using the `ProcedureCatalog`.

I've removed or `@Ignore`d the tests against `SparkSessionCatalog` in this PR.